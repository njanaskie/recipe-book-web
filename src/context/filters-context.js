import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    const filterIngredients = (ingredients) => {
        filtersDispatch({ type: 'SET_INGREDIENT_FILTER', ingredients })
    }

    const filterCuisine = (cuisine) => {
        filtersDispatch({ type: 'SET_CUISINE_FILTER', cuisine })
    }

    const filterType = (recipeType) => {
        filtersDispatch({ type: 'SET_RECIPE_TYPE_FILTER', recipeType })
    }

    const filterCustomTags = (customTags) => {
        filtersDispatch({ type: 'SET_CUSTOM_TAG_FILTER', customTags })
    }

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch, filterCuisine, filterIngredients, filterType, filterCustomTags }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }