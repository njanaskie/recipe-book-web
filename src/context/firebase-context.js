import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../firebase/firebase';
import { history } from '../routers/AppRouter';
import authReducer from '../reducers/auth'
import LoadingPage from '../components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/home')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, loading, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }