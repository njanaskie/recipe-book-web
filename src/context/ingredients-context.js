import React, { useContext, useReducer } from 'react';
import ingredientsReducer from '../reducers/ingredients'
import database from '../firebase/firebase'

const IngredientsContext = React.createContext()

export const useIngredientsContext = () => useContext(IngredientsContext)

const IngredientsProvider = ({ children }) => {
    const [ingredients, dispatch ] = useReducer(ingredientsReducer, [])

    const removeIngredient = ({ id } = {}) => {
        return database.collection('ingredients').doc(id).delete().then(() => {
            dispatch({ type: 'REMOVE_INGREDIENT', id })
        });
    };

    const addIngredient = (ingredient) => {
        database.collection('ingredients').add(ingredient).then((ref) => {
            dispatch(({ type: 'ADD_INGREDIENT', ingredient: {id: ref.key, ...ingredient} }))
        })
    }

    return (
        <IngredientsContext.Provider value={{ ingredients, dispatch, removeIngredient, addIngredient }}>
            {children}
        </IngredientsContext.Provider>
    )
}

export { IngredientsProvider, IngredientsContext as default }