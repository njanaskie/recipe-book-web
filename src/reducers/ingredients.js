
// Ingredients Reducer

const ingredientsReducerDefaultState = [];

const ingredientsReducer = (state = ingredientsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_INGREDIENT':
            return [
                ...state,
                action.ingredient
            ];
        case 'REMOVE_INGREDIENT':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_INGREDIENT':
            return state.map((ingredient) => {
                if (ingredient.id === action.id) {
                    return {
                        ...ingredient,
                        ...action.updates
                    };
                } else {
                    return ingredient;
                };
            });
        case 'SET_INGREDIENTS':
            return action.ingredients;
        default:
            return state;
    }
}

export default ingredientsReducer;