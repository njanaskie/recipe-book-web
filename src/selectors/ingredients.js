
export default (ingredients) => {
    return ingredients
    .sort((a, b) => {
        if (a.category === b.category) {
            return b.category - a.category
        }
         return a.category > b.category ? 1 : -1
    })
}