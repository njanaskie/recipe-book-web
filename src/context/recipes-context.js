import React, { useContext, useReducer } from 'react';
import recipesReducer from '../reducers/recipes'
import database from '../firebase/firebase'
import uuid from 'uuid'
import { useFirebaseContext } from './firebase-context'


const RecipesContext = React.createContext()

export const useRecipesContext = () => useContext(RecipesContext)

const RecipesProvider = ({ children }) => {
    const { user } = useFirebaseContext() 
    const [recipes, recipeDispatch] = useReducer(recipesReducer, [])
    const customId = uuid()

    const addRecipe = (recipe) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(customId).set(recipe).then(() => {
            recipeDispatch({ type: 'ADD_RECIPE', recipe: {id: customId, ...recipe} })
        })
    }

    const editRecipe = (id, updates) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).update(updates).then(() => {
            recipeDispatch({ type: 'EDIT_RECIPE', id, updates })
        })
    }

    const removeRecipe = (id) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).delete().then(() => {
            recipeDispatch({ type: 'REMOVE_RECIPE', id })
        })
    }

    return (
        <RecipesContext.Provider value={{ recipes, recipeDispatch, addRecipe, editRecipe, removeRecipe }}>
            {children}
        </RecipesContext.Provider>
    )
}

export { RecipesProvider, RecipesContext as default }