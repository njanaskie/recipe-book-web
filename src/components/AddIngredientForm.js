import React from 'react'
import { Form } from 'semantic-ui-react'
import { useIngredientsContext } from '../context/ingredients-context'
import { addIngredientService } from '../services/ingredientServices'

const AddIngredientForm = (props) => {
    const initialFormState = {
        name: '',
        category: '',
        price: '',
        error: ''
    }
    const [state, setState] = React.useState(initialFormState)
    const { dispatch, addIngredient } = useIngredientsContext()

    const handleAddIngredient = (e) => {
        e.preventDefault()
        const ingredient = {
            name: state.name,
            category: state.category,
            price: state.price
        }

        if (!state.name || !state.category || !state.price) {
            setState({ error: 'Please provide name, category, and price'})
        } else {
            // addIngredient(ingredient)
            addIngredientService(ingredient, dispatch)
            dispatch({ type: 'ADD_INGREDIENT', ingredient })
            setState(initialFormState)
        }
    }

    const onNameChange = (e) => {
        const value = e.target.value
        setState({ ...state, name: value })
    }

    const onCategoryChange = (e, result) => {
        const { value } = result || e.target
        setState({ ...state, category: value })
    }

    const onPriceChange = (e, result) => {
        const { value } = result || e.target
        setState({ ...state, price: value })
    }

    const categoryOptions = [
        {key: 'dairy', text: 'Dairy', value: 'dairy'},
        {key: 'meats', text: 'Meats', value: 'meats'},
        {key: 'seafood', text: 'Seafood', value: 'seafood'},
        {key: 'vegetables', text: 'Vegetables', value: 'vegetables'},
        {key: 'baking and grains', text: 'Baking and Grains', value: 'baking and grains'},
        {key: 'added sweeteners', text: 'Added Sweeteners', value: 'added sweeteners'},
        {key: 'condiments', text: 'Condiments', value: 'condiments'},
        {key: 'spices', text: 'Spices', value: 'spices'},
        {key: 'fish', text: 'Fish', value: 'fish'},
        {key: 'fruits', text: 'Fruits', value: 'fruits'},
        {key: 'oils', text: 'Oils', value: 'oils'},
        {key: 'seasonings', text: 'Seasonings', value: 'seasonings'},
        {key: 'sauces', text: 'Sauces', value: 'sauces'},
        {key: 'legumes', text: 'Legumes', value: 'legumes'},
        {key: 'alcohol', text: 'Alcohol', value: 'alcohol'},
        {key: 'soup', text: 'Soup', value: 'soup'},
        {key: 'nuts', text: 'Nuts', value: 'nuts'},
        {key: 'dairy alternatives', text: 'Dairy Alternatives', value: 'dairy alternatives'},
        {key: 'desserts and snacks', text: 'Desserts and Snacks', value: 'desserts and snacks'},
        {key: 'beverages', text: 'Beverages', value: 'beverages'},
    ]

    const priceOptions = [
        {key: 1, text: 1, value: 1},
        {key: 2, text: 2, value: 2},
        {key: 3, text: 3, value: 3},
    ]

    return (
        <div className='content-container'>
            <div className='form-container form-title'>
                <h4>Add an ingredient to the site</h4>
            </div>
                <Form onSubmit={handleAddIngredient} className='form-container'>
                    {state.error && <p className='form__error'>{state.error}</p>}
                    <Form.Group widths='equal' inline>
                            <Form.Input
                                id='set-name'
                                fluid
                                type='text'
                                placeholder='Name'
                                value={state.name || ''}
                                onChange={onNameChange}
                            />
                            <Form.Dropdown
                                id='set-category'
                                placeholder='Category'
                                fluid selection
                                clearable
                                value={state.category || ''}
                                onChange={onCategoryChange}
                                options={categoryOptions.sort((a,b) => a.key.localeCompare(b.key)).map(category => category)}
                            />
                            <Form.Dropdown
                                id='set-price'
                                placeholder='Price'
                                fluid selection
                                clearable
                                value={state.price || ''}
                                onChange={onPriceChange}
                                options={priceOptions.map(price => price)}
                            />
                            <Form.Button fluid>Add Ingredient</Form.Button>
                        </Form.Group>
                </Form>
        </div>

    )
}

export default AddIngredientForm