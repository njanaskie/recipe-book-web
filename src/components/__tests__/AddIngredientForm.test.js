import React from 'react'
import { shallow, mount } from 'enzyme'
import AddIngredientForm from '../AddIngredientForm'
import ingredients from '../../../tests/fixtures/ingredients'
import * as IngredientsContext from '../../../context/ingredients-context'

let dispatch, setStateSpy, setName, setCategory, setPrice, setError, addIngredient;

beforeEach(() => {
    dispatch = jest.fn()
    setStateSpy = jest.spyOn(React, 'useState')
    setName = jest.fn()
    setCategory = jest.fn()
    setPrice = jest.fn()
    setError = jest.fn()
    addIngredient = jest.fn()
})

afterEach(() => {
    jest.clearAllMocks();
});

test('should render AddIngredientForm correctly',() => {
    jest.spyOn(IngredientsContext, 'useIngredientsContext').mockImplementation(() => dispatch)
    const wrapper = shallow(<AddIngredientForm />)
    expect(wrapper).toMatchSnapshot()
});

test('should render error message for invalid form submission', () => {
    const wrapper = shallow(<AddIngredientForm />)
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(wrapper).toMatchSnapshot()
})

test('should set error for invalid form submission', () => {
    setStateSpy.mockImplementation((init) => [init, setError]);
    const wrapper = shallow(<AddIngredientForm />)
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(setError).toHaveBeenCalled()
})

test('should set name on input change', () => {
    const value = 'New name'
    setStateSpy.mockImplementation((init) => [init, setName]);
    const wrapper = shallow(<AddIngredientForm />)
    wrapper.find('#set-name').props().onChange({ target: { value }});
    expect(setName).toHaveBeenLastCalledWith(value)
});

test('should set category on select change', () => {
    const value = 'New category'
    setStateSpy.mockImplementation((init) => [init, setCategory]);
    const wrapper = shallow(<AddIngredientForm />)
    wrapper.find('#set-category').props().onChange({ target: { value }});
    expect(setCategory).toHaveBeenCalledWith(value)
});

test('should set price on select change', () => {
    const value = 1
    setStateSpy.mockImplementation((init) => [init, setPrice]);
    const wrapper = shallow(<AddIngredientForm />)
    wrapper.find('#set-price').props().onChange({ target: { value }});
    expect(setPrice).toHaveBeenCalledWith(value)
});

// refactor to use one state, currently submitting without ingredient
test('should call onSubmit prop for valid form submission', () => {
    setStateSpy.mockImplementation((init) => [init, addIngredient]);
    const wrapper = shallow(<AddIngredientForm ingredient={ingredients[0]} onSubmit={addIngredient} />)
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    })
    // expect(wrapper.state('error')).toBe('')
    // expect(addIngredient).toHaveBeenLastCalledWith({
    //     name: ingredients[0].name,
    //     category: ingredients[0].category,
    //     price: ingredients[0].price
    // })
    expect(addIngredient).toHaveBeenCalled()
})