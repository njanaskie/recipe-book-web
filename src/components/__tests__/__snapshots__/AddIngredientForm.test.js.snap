// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddIngredientForm correctly 1`] = `
<div>
  <p>
    Add Ingredient
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      id="set-name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <select
      id="set-category"
      onChange={[Function]}
      value="dairy"
    >
      <option
        value="dairy"
      >
        Dairy
      </option>
      <option
        value="meats"
      >
        Meats
      </option>
      <option
        value="seafood"
      >
        Seafood
      </option>
      <option
        value="vegetables"
      >
        Vegetables
      </option>
      <option
        value="fruit"
      >
        Fruit
      </option>
      <option
        value="baking and grains"
      >
        Baking and Grains
      </option>
      <option
        value="added sweeteners"
      >
        Added Sweeteners
      </option>
      <option
        value="condiments"
      >
        Condiments
      </option>
      <option
        value="spices"
      >
        Spices
      </option>
      <option
        value="fish"
      >
        Fish
      </option>
      <option
        value="oils"
      >
        Oils
      </option>
      <option
        value="seasonings"
      >
        Seasonings
      </option>
      <option
        value="sauces"
      >
        Sauces
      </option>
      <option
        value="legumes"
      >
        Legumes
      </option>
      <option
        value="alcohol"
      >
        Alcohol
      </option>
      <option
        value="soup"
      >
        Soup
      </option>
      <option
        value="nuts"
      >
        Nuts
      </option>
      <option
        value="dairy alternatives"
      >
        Dairy Alternatives
      </option>
      <option
        value="desserts and snacks"
      >
        Desserts and Snacks
      </option>
      <option
        value="beverages"
      >
        Beverages
      </option>
    </select>
    <select
      id="set-price"
      onChange={[Function]}
      value="1"
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="2"
      >
        2
      </option>
      <option
        value="3"
      >
        3
      </option>
    </select>
    <button>
      Add Ingredient
    </button>
  </form>
</div>
`;

exports[`should render error message for invalid form submission 1`] = `
<div>
  <p>
    Add Ingredient
  </p>
  <form
    onSubmit={[Function]}
  >
    <p>
      Please provide name, category, and price
    </p>
    <input
      autoFocus={true}
      id="set-name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <select
      id="set-category"
      onChange={[Function]}
      value="dairy"
    >
      <option
        value="dairy"
      >
        Dairy
      </option>
      <option
        value="meats"
      >
        Meats
      </option>
      <option
        value="seafood"
      >
        Seafood
      </option>
      <option
        value="vegetables"
      >
        Vegetables
      </option>
      <option
        value="fruit"
      >
        Fruit
      </option>
      <option
        value="baking and grains"
      >
        Baking and Grains
      </option>
      <option
        value="added sweeteners"
      >
        Added Sweeteners
      </option>
      <option
        value="condiments"
      >
        Condiments
      </option>
      <option
        value="spices"
      >
        Spices
      </option>
      <option
        value="fish"
      >
        Fish
      </option>
      <option
        value="oils"
      >
        Oils
      </option>
      <option
        value="seasonings"
      >
        Seasonings
      </option>
      <option
        value="sauces"
      >
        Sauces
      </option>
      <option
        value="legumes"
      >
        Legumes
      </option>
      <option
        value="alcohol"
      >
        Alcohol
      </option>
      <option
        value="soup"
      >
        Soup
      </option>
      <option
        value="nuts"
      >
        Nuts
      </option>
      <option
        value="dairy alternatives"
      >
        Dairy Alternatives
      </option>
      <option
        value="desserts and snacks"
      >
        Desserts and Snacks
      </option>
      <option
        value="beverages"
      >
        Beverages
      </option>
    </select>
    <select
      id="set-price"
      onChange={[Function]}
      value="1"
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="2"
      >
        2
      </option>
      <option
        value="3"
      >
        3
      </option>
    </select>
    <button>
      Add Ingredient
    </button>
  </form>
</div>
`;
